packages/widget.client.unstable/src/request.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { HostMessage, WidgetMessage } from "@osdk/widget-api.unstable";
+
+export function serializeRequest(
+  id: string,
+  input: RequestInfo | URL,
+  init: RequestInit | undefined,
+): WidgetMessage.Payload._unstable_FetchRequest {
+  if (typeof input !== "string") {
+    throw new Error("Only string input URLs are supported");
+  }
+  const body = init?.body;
+  if (body !== undefined && typeof body !== "string") {
+    throw new Error("Only string request body values are supported");
+  }
+  return {
+    id,
+    url: input,
+    method: init?.method ?? "GET",
+    headers: serializeHeaders(init?.headers),
+    body: body as any,
+  };
+}
+
+function serializeHeaders(headers: HeadersInit | undefined) {
+  if (headers === undefined || !(headers instanceof Headers)) {
```

Comment:
Apologies I know this is copied from me, but for completeness we could also error if not an instance of Headers, at least then all of the data type assumptions of this code error loudly if the client did something we don't support

Reply:
So `HeadersInit` can be any of the following `string, string][] | Record<string, string> | Headers`. But also you can pass all of those to `new Headers()` to ensure you have a headers object. 

You could also therefore simplify serialization to just `[...new Headers(headers ?? {}).entries()]` and rehydrate on the other side with `new Headers(...)`